test/e2e/tests/single_zone_e2e_test.go

```diff
@@ -159,6 +161,147 @@ var _ = Describe("GCE PD CSI Driver", func() {
+		defer func() {
+			// Detach Disk
+			err = client.ControllerUnpublishVolume(volIDA, instance.GetNodeID())
+			if err != nil {
+				klog.Errorf("Failed to detach disk: %v", err)
+			}
+		}()
+
+		// Setup: Create Disk B
+		volNameB, volIDB := createAndValidateUniqueZonalDisk(client, p, z)
+		defer func() {
+			// Delete Disk
+			err := client.DeleteVolume(volIDB)
+			Expect(err).To(BeNil(), "DeleteVolume failed")
+
+			// Validate Disk Deleted
+			_, err = computeService.Disks.Get(p, z, volNameB).Do()
+			Expect(gce.IsGCEError(err, "notFound")).To(BeTrue(), "Expected disk to not be found")
+		}()
+
+		// Setup: Attach Disk B
+		err = client.ControllerPublishVolume(volIDB, instance.GetNodeID())
+		Expect(err).To(BeNil(), "ControllerPublishVolume failed with error for disk %v on node %v: %v", volIDB, instance.GetNodeID())
+		defer func() {
+			// Detach Disk
+			err = client.ControllerUnpublishVolume(volIDB, instance.GetNodeID())
+			if err != nil {
+				klog.Errorf("Failed to detach disk: %v", err)
+			}
+		}()
+
+		// Setup: Stage Disk A
+		stageDirA := filepath.Join("/tmp/", volNameA, "stage")
+		err = client.NodeStageExt4Volume(volIDA, stageDirA)
+		Expect(err).To(BeNil(), "failed to stage volume")
+
+		// Assert: Stage to same location with different fstype should fail
+		err = client.NodeStageVolume(volIDA, stageDirA, &csi.VolumeCapability{
+			AccessType: &csi.VolumeCapability_Mount{
+				Mount: &csi.VolumeCapability_MountVolume{
+					FsType: "xfs",
+				},
+			},
+			AccessMode: &csi.VolumeCapability_AccessMode{
+				Mode: csi.VolumeCapability_AccessMode_SINGLE_NODE_WRITER,
+			},
+		})
+		e, ok := status.FromError(err)
+		Expect(ok).To(BeTrue(), "Could not get error type from err", err)
+		Expect(e.Code()).To(Equal(codes.AlreadyExists), "Volume staged with different fs type should result in already exists error")
+
+		// Assert: Stage to same location with same fstype should work
+		err = client.NodeStageVolume(volIDA, stageDirA, &csi.VolumeCapability{
+			AccessType: &csi.VolumeCapability_Mount{
+				Mount: &csi.VolumeCapability_MountVolume{
+					FsType: "ext4",
+				},
+			},
+			AccessMode: &csi.VolumeCapability_AccessMode{
+				Mode: csi.VolumeCapability_AccessMode_SINGLE_NODE_WRITER,
+			},
+		})
+		Expect(err).To(BeNil(), "Staged volume to same location with same fs type should work")
+
+		// Assert: Stage volume using block to location with fs type already should always work
+		err = client.NodeStageBlockVolume(volIDA, stageDirA)
+		Expect(err).To(BeNil(), "Staged volume of block type should always work")
+
+		defer func() {
+			// Unstage Disk
+			err = client.NodeUnstageVolume(volIDA, stageDirA)
+			if err != nil {
+				klog.Errorf("Failed to unstage volume: %v", err)
+			}
+			fp := filepath.Join("/tmp/", volNameA)
+			err = testutils.RmAll(instance, fp)
+			if err != nil {
+				klog.Errorf("Failed to rm file path %s: %v", fp, err)
+			}
+		}()
+
+		// Assert: Stage Disk B to Disk A position and fail even both as EXT4
```

Comment:
Can we run each scenario as a separate test case?

Reply:
There's a lot of boilerplate setup for "later assertions" as asserting things for publish depends on doing a attach and stage. I actually think this is a lot cleaner especially since i've documented it quite thoroughly. In the passing case we're happy, in the failing case I'm not really concerned about "knock on errors" since a failure is a failure and you can debug based on the first error seen.